src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,228 @@
+    {
+      // TODO: Do we need to check if the session id is already in use?
+      var resourceReaper = new ResourceReaper(sessionId);
+
+      try
+      {
+        await resourceReaper.StartAsync(ct)
+          .ConfigureAwait(false);
+        return resourceReaper;
+      }
+      catch (Exception)
+      {
+        await resourceReaper.DisposeAsync()
+          .ConfigureAwait(false);
+        throw;
+      }
+    }
+
+    public ValueTask DisposeAsync()
+    {
+      if (this.disposed)
+      {
+        return default;
+      }
+
+      this.disposed = true;
+
+      this.cancellationTokenSource.Cancel();
+      this.cancellationTokenSource.Dispose();
+
+      return this.resourceReaperContainer?.DisposeAsync() ?? default;
+    }
+
+    private async Task StartAsync(CancellationToken ct = default)
+    {
+      var ryukInitializedTaskSource = new TaskCompletionSource<bool>();
+
+      await this.resourceReaperContainer.StartAsync(ct)
+        .ConfigureAwait(false);
+
+      _ = this.MaintainRyukConnection(ryukInitializedTaskSource, ct);
+
+      await ryukInitializedTaskSource.Task
+        .ConfigureAwait(false);
+    }
+
+    private async Task MaintainRyukConnection(TaskCompletionSource<bool> ryukInitializedTaskSource, CancellationToken ct = default)
+    {
+      using (var initTimeoutCancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(10)))
+      {
+        using (var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.cancellationTokenSource.Token, initTimeoutCancellationTokenSource.Token, ct))
+        {
+          var endpoint = new IPEndPoint(IPAddress.Loopback, this.resourceReaperContainer.GetMappedPublicPort(RyukPort));
+
+          using (var tcpClient = new TcpClient())
+          {
+            while (!linkedCancellationTokenSource.IsCancellationRequested)
+            {
+              try
+              {
+                tcpClient.Connect(endpoint);
+
+                var stream = tcpClient.GetStream();
+
+                var filter = $"label={ResourceReaperSessionLabel}={this.SessionId:D}\n";
+
+                var sendBytes = Encoding.ASCII.GetBytes(filter);
+
+                var readBytes = new byte[4];
+
+                if (!ryukInitializedTaskSource.Task.IsCompleted)
+                {
+                  bool hasAcknowledge;
+
+                  await stream.WriteAsync(sendBytes, 0, sendBytes.Length, linkedCancellationTokenSource.Token)
+                    .ConfigureAwait(false);
+
+                  await stream.FlushAsync(linkedCancellationTokenSource.Token)
+                    .ConfigureAwait(false);
+
+                  do
+                  {
+                    _ = await stream.ReadAsync(readBytes, 0, readBytes.Length, linkedCancellationTokenSource.Token)
+                      .ConfigureAwait(false);
+                    hasAcknowledge = "ack\n".Equals(Encoding.ASCII.GetString(readBytes), StringComparison.OrdinalIgnoreCase);
```

Comment:
I think it would be better to keep reading the stream line by line as this would break if ryuk sends some other line message before the ack:
Example:
```
hello\n
ack\n
```

This code would read the stream as `hell`, `o\nac`, `k\n...`. Therefore the `ResourceReaper` would never initialize successfully although ryuk has responded with `ack\n`.

Reply:
👍 I would prefere `ReadLineAsync` as well, but there is no overload that includes the `CancellationToken` as far as I know. I'll take a look at it again.

Reply:
I changed the part that reads from the stream. It should handle different messages now too, but does that happen at all:
https://github.com/testcontainers/moby-ryuk/blob/a0508338f5a0d2f4f8de0dc5ff7355f32bf0c8ca/main.go#L125?

Reply:
In the current implementation of Ryuk this can not happen as far as I known, but I am not 100% sure that this can not change in the future.

You are right about that missing CancellationToken parameter in `ReadLineAsync`. I overlooked that.